File: typescript/src/base/ValueRegistry.ts
import { Decimal } from 'decimal.js';
import { Value, FormulaValue, ConditionValue, ResourceValue } from './Value';
import { ExpressionParser, ParseResult } from './ExpressionParser';

export class ValueRegistry {
  private values: Map<string, Value> = new Map();
  private dependencies: Map<string, Set<string>> = new Map();

  get Values(): Map<string, Value> {
    return this.values;
  }

  private getDependencies(valueId: string): Set<string> {
    let deps = this.dependencies.get(valueId);
    if (!deps) {
      deps = new Set();
      this.dependencies.set(valueId, deps);
    }
    return deps;
  }

  getWrappedValue(valueId: string): Decimal {
    return this.get(valueId).getValue(this);
  }

  get(valueId: string): Value {
    const value = this.values.get(valueId);
    if (!value) {
      throw new Error(`Value with ID '${valueId}' not found in registry`);
    }
    return value;
  }

  getTyped<T extends Value>(valueId: string): T {
    const value = this.values.get(valueId);
    if (!value) {
      throw new Error(`Value with ID '${valueId}' not found in registry`);
    }
    return value as T;
  }

  addFormula(valueId: string, expression: string): FormulaValue {
    const existingValue = this.values.get(valueId);
    if (existingValue) {
      console.log(`Value ${valueId} already added`);
      return existingValue as FormulaValue;
    }

    const result = ExpressionParser.parseWithDependencies(expression);
    const formulaValue = new FormulaValue(valueId, result.function);
    console.log(
      `Adding formula ${valueId} with dependencies ${Array.from(result.dependencies).join(', ')}`
    );

    for (const dependency of result.dependencies) {
      this.getDependencies(dependency).add(valueId);
    }
    this.add(formulaValue);
    return formulaValue;
  }

  addCondition(
    valueId: string,
    expression: string,
    isMet = false,
    onConditionMet?: (valueRegistry: ValueRegistry) => void
  ): ConditionValue {
    if (this.values.has(valueId)) {
      // Remove old value since condition value has state
      this.values.delete(valueId);
    }

    const result = ExpressionParser.parseWithDependencies(expression);
    const conditionValue = new ConditionValue(
      valueId,
      result.function,
      isMet,
      onConditionMet
    );
    console.log(
      `Adding condition ${valueId} with dependencies ${Array.from(result.dependencies).join(', ')}`
    );

    for (const dependency of result.dependencies) {
      this.getDependencies(dependency).add(valueId);
    }
    this.add(conditionValue);

    return conditionValue;
  }

  addResource(
    valueId: string,
    deltaExpression = '0',
    baseExpression = '0',
    multExpression = '1',
    flatExpression = '0'
  ): ResourceValue {
    console.log(`Adding resource ${valueId}`);
    this.addFormula(`${valueId}_delta`, deltaExpression);
    this.addFormula(`${valueId}_base`, baseExpression);
    this.addFormula(`${valueId}_mult`, multExpression);
    this.addFormula(`${valueId}_flat`, flatExpression);
    this.addFormula(
      `${valueId}_calculated`,
      `${valueId}_base * ${valueId}_mult + ${valueId}_flat`
    );

    const resource = new ResourceValue(valueId);
    this.getDependencies(`${valueId}_calculated`).add(valueId);
    this.add(resource);

    return resource;
  }

  add(value: Value): void {
    if (!this.values.has(value.valueId)) {
      this.values.set(value.valueId, value);
    }
  }

  printOutDependencies(): void {
    for (const [key, value] of this.dependencies) {
      console.log(`${key} -> ${Array.from(value).join(', ')}`);
    }
  }

  invalidate(value: Value): void {
    for (const dependency of this.getDependencies(value.valueId)) {
      this.values.get(dependency)?.invalidate(this);
    }
  }

  update(tick: number): void {
    for (const value of this.values.values()) {
      if (value instanceof ResourceValue) {
        value.update(this, tick);
      }
    }
  }

  remove(valueId: string): boolean {
    if (!this.values.has(valueId)) {
      return false;
    }

    // Check dependencies
    const deps = this.dependencies.get(valueId);
    if (deps && deps.size > 0) {
      console.log(`Cannot remove ${valueId} because other values depend on it`);
      return false;
    }

    // Remove from other values' dependencies
    for (const deps of this.dependencies.values()) {
      deps.delete(valueId);
    }

    // Remove value
    this.values.delete(valueId);

    return true;
  }
} 

File: BaseSystem/ValueRegistry.cs
using System;
using System.Collections.Generic;
using BreakInfinity;

namespace BaseSystem
{
    public class ValueRegistry
    {
        private Dictionary<string, Value> values = new();
        public Dictionary<string, Value> Values => values;

        private Dictionary<string, HashSet<string>> dependencies = new();

        private HashSet<string> GetDependencies(string valueId)
        {
            if (!dependencies.TryGetValue(valueId, out var deps))
            {
                deps = new HashSet<string>();
                dependencies[valueId] = deps;
            }
            return deps;
        }

        public BigDouble GetWrapedValue(string valueId)
        {
            return Get(valueId).GetValue(this);
        }

        public Value Get(string valueId)
        {
            if (!values.TryGetValue(valueId, out var value))
            {
                throw new KeyNotFoundException($"Value with ID '{valueId}' not found in registry");
            }
            return value;
        }

        public T Get<T>(string valueId)
            where T : Value
        {
            if (!values.TryGetValue(valueId, out var value))
            {
                throw new KeyNotFoundException($"Value with ID '{valueId}' not found in registry");
            }
            return (T)value;
        }

        public FormulaValue AddFormula(string valueId, string expression)
        {
            if (values.TryGetValue(valueId, out Value value))
            {
                Console.WriteLine($"Value {valueId} already added");
                return value as FormulaValue;
            }

            var result = ExpressionParser.ParseWithDependencies(expression);
            var formulaValue = new FormulaValue(valueId, result.Function);
            Console.WriteLine(
                $"Adding formula {valueId} with dependencies {string.Join(", ", result.Dependencies)}"
            );

            foreach (var dependency in result.Dependencies)
                GetDependencies(dependency).Add(valueId);
            Add(formulaValue);
            return formulaValue;
        }

        public ConditionValue AddCondition(
            string valueId,
            string expression,
            // This value comes from saved data
            bool isMet = false,
            Action<ValueRegistry> onConditionMet = null
        )
        {
            if (values.TryGetValue(valueId, out Value value))
            {
                // Remove old value since condition value has state
                values.Remove(valueId);
            }

            var result = ExpressionParser.ParseWithDependencies(expression);
            var conditionValue = new ConditionValue(
                valueId,
                result.Function,
                isMet,
                onConditionMet
            );
            Console.WriteLine(
                $"Adding condition {valueId} with dependencies {string.Join(", ", result.Dependencies)}"
            );

            foreach (var dependency in result.Dependencies)
                GetDependencies(dependency).Add(valueId);
            Add(conditionValue);

            return conditionValue;
        }

        public ResourceValue AddResource(
            string valueId,
            string deltaExpression = "0",
            string baseExpression = "0",
            string multExpression = "1",
            string flatExpression = "0"
        )
        {
            Console.WriteLine($"Adding resource {valueId}");
            AddFormula($"{valueId}_delta", deltaExpression);
            AddFormula($"{valueId}_base", baseExpression);
            AddFormula($"{valueId}_mult", multExpression);
            AddFormula($"{valueId}_flat", flatExpression);
            AddFormula(
                $"{valueId}_calculated",
                $"{valueId}_base * {valueId}_mult + {valueId}_flat"
            );

            var resource = new ResourceValue(valueId);
            GetDependencies($"{valueId}_calculated").Add(valueId);
            Add(resource);

            return resource;
        }

        public void Add(Value value)
        {
            if (!values.ContainsKey(value.ValueId))
            {
                values[value.ValueId] = value;
            }
        }

        public void PrintOutDependencies()
        {
            foreach (var dependency in dependencies)
            {
                Console.WriteLine($"{dependency.Key} -> {string.Join(", ", dependency.Value)}");
            }
        }

        public void Invalidate(Value value)
        {
            foreach (var dependency in GetDependencies(value.ValueId))
            {
                values[dependency].Invalidate(this);
            }
        }

        public void Update(float tick)
        {
            foreach (var value in values.Values)
            {
                if (value is ResourceValue resource)
                {
                    resource.Update(this, tick);
                }
            }
        }

        public bool Remove(string valueId)
        {
            if (!values.ContainsKey(valueId))
                return false;

            // 의존성 정리
            if (dependencies.ContainsKey(valueId) && dependencies[valueId].Count > 0)
            {
                Console.WriteLine($"Cannot remove {valueId} because other values depend on it");
                return false;
            }

            // 다른 값들의 의존성에서도 제거
            foreach (var deps in dependencies.Values)
            {
                deps.Remove(valueId);
            }

            // 값 제거
            values.Remove(valueId);

            return true;
        }
    }
}
